controllers/release/release_adapter.go

```diff
@@ -0,0 +1,306 @@
+		}
+
+		pipelineRun, err = a.createReleasePipelineRun(releaseStrategy)
+		if err != nil {
+			return results.RequeueWithError(err)
+		}
+
+		a.logger.Info("Created release PipelineRun",
+			"PipelineRun.Name", pipelineRun.Name, "PipelineRun.Namespace", pipelineRun.Namespace)
+	}
+
+	if !a.release.HasStarted() {
+		a.release.Status.ReleasePipelineRun = fmt.Sprintf("%s%c%s",
+			pipelineRun.Namespace, types.Separator, pipelineRun.Name)
+		a.release.Status.MarkRunning()
+
+		return results.RequeueOnErrorOrContinue(a.updateStatus())
+	}
+
+	return results.ContinueProcessing()
+}
+
+// EnsureReleasePipelineStatusIsTracked is an operations that will ensure that the release PipelineRun status is tracked
+// in the Release being processed.
+func (a *Adapter) EnsureReleasePipelineStatusIsTracked() (results.OperationResult, error) {
+	if !a.release.HasStarted() || a.release.IsDone() {
+		return results.ContinueProcessing()
+	}
+
+	pipelineRun, err := a.getReleasePipelineRun()
+	if err != nil {
+		return results.RequeueWithError(err)
+	}
+	if pipelineRun != nil {
+		return results.RequeueOnErrorOrContinue(a.trackReleasePipelineRunStatus(pipelineRun))
+	}
+
+	return results.ContinueProcessing()
+}
+
+// createReleasePipelineRun creates and returns a new release PipelineRun. The new PipelineRun will include owner
+// annotations, so it triggers Release reconciles whenever it changes.
+func (a *Adapter) createReleasePipelineRun(releaseStrategy *v1alpha1.ReleaseStrategy) (*v1beta1.PipelineRun, error) {
+	pipelineRun := tekton.CreatePipelineRunFromReleaseStrategy(releaseStrategy, a.release)
+	err := libhandler.SetOwnerAnnotations(a.release, pipelineRun)
+	if err != nil {
+		return nil, err
+	}
+	err = a.client.Create(a.context, pipelineRun)
+	if err != nil {
+		return nil, err
+	}
+
+	return pipelineRun, nil
+}
+
+// finalizeRelease will finalize the Release being processed, removing the associated resources.
+func (a *Adapter) finalizeRelease() error {
+	releasePipelineRun, err := a.getReleasePipelineRun()
+	if err != nil {
+		return err
+	}
+
+	if releasePipelineRun != nil {
+		err = a.client.Delete(a.context, releasePipelineRun)
+		if err != nil {
+			return err
+		}
+	}
+
+	a.logger.Info("Successfully finalized Release")
+
+	return nil
+}
+
+// getReleaseLink returns the ReleaseLink referenced by the Release being processed. If the ReleaseLink is not found or
+// the Get operation failed, an error will be returned.
+func (a *Adapter) getReleaseLink() (*v1alpha1.ReleaseLink, error) {
+	releaseLink := &v1alpha1.ReleaseLink{}
+	err := a.client.Get(a.context, types.NamespacedName{
+		Namespace: a.release.Namespace,
+		Name:      a.release.Spec.ReleaseLink,
+	}, releaseLink)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return releaseLink, nil
+}
+
+// getReleasePipelineRun returns the PipelineRun referenced by the Release being processed. If the PipelineRun is not
+// found or the List operation failed, an error will be returned.
```

Comment:
if it is not found but the list operation succeeds, we return `nil,nil` not an error